cpp/src/cylon/repartition.hpp

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <numeric>
+#include <cylon/util/macros.hpp>
+
+namespace cylon {
+
+/**
+ * Divide the rows evenly among all workers
+ * number of workers = row_counts.size()
+ * If the rows can not be divided evenly among all workers,
+ * first k workers get one extra row.
+ * For example:
+ *   row_counts{0, 14, 0, 0}
+ *   Result: {4, 4, 3, 3}
+ * @param row_counts
+ * @return
+ */
+template<typename T>
+std::vector<T> DivideRowsEvenly(const std::vector<T> &row_counts) {
+  auto nworkers = row_counts.size();
+  auto all_rows = std::accumulate(row_counts.begin(), row_counts.end(), 0);
+  T rows_per_worker = all_rows / nworkers;
+  T workers_with_extra_row = all_rows % nworkers;
+  std::vector<T> even_rows(nworkers, rows_per_worker);
+  std::fill(even_rows.begin(), even_rows.begin() + workers_with_extra_row, rows_per_worker + 1);
+  return even_rows;
+}
+
+/**
+ * determine the target worker of the first row
+ * calculate the global order of the first row of that worker
+ * @param target_row_counts
+ * @param my_row_start
+ * @return
+ */
+template<typename T>
+std::pair<T, T> FirstTarget(const std::vector<T> &target_row_counts, T my_row_start) {
+  // target worker of my first row
+  T first_target_rank = -1;
+  // global order of the first row of the first target worker
+  T first_target_start_row = 0;
+  for (int i = 0; i < target_row_counts.size(); ++i) {
+    if (first_target_start_row + target_row_counts[i] > my_row_start) {
+      first_target_rank = i;
+      break;
+    } else {
+      first_target_start_row += target_row_counts[i];
+    }
+  }
+
+  return std::make_pair(first_target_rank, first_target_start_row);
+}
+
+/**
+ * calculates indicies for all-to-all operation for a worker
+ * given the current row_counts and requested row counts
+ * each worker calls this function to determine ranges of rows to send to each worker
+ * In the returned vector:
+ *   rows in the range of to_all[0] - to_all[1] shall be send to worker 0
+ *   rows in the range of to_all[1] - to_all[2] shall be send to worker 2
+ *   ...
+ *   rows in the range of to_all[size-2] - to_all[size-1] shall be send to the last worker
+ *
+ * make sure:
+ *   number_of_workers == current_row_counts.size() == target_row_counts.size()
+ * make sure:
+ *   sum_of(current_row_counts) == sum_of(target_row_counts)
+ * we do not check those in the function
+ * @param my_rank
+ * @param current_row_counts
+ * @param target_row_counts
+ * @return
+ */
+template<typename T>
+std::vector<T> RowIndicesToAll(int32_t my_rank,
+                               const std::vector<T> &current_row_counts,
+                               const std::vector<T> &target_row_counts) {
```

Comment:
I have a feeling that calculating prefix sum might help, and relax some of the computations of the util methods as well, IMO.

Reply:
i could not think of an improvement with a prefix sum.